File: android/src/main/kotlin/io/sentient/android/chat/message/SendAnchor.kt
package io.sentient.android.chat.message

/** Pure state for the one-shot viewport movement caused by an outbound user row. */
internal data class SendAnchorState(val observed: Set<String> = emptySet())

/** Returns the first newly observed send identity, or null for existing/assistant rows. */
internal fun reduceSendAnchor(
    state: SendAnchorState,
    identities: Set<String>,
): Pair<SendAnchorState, String?> {
    val newIdentity = identities.firstOrNull { it !in state.observed }
    return SendAnchorState(state.observed + identities) to newIdentity
}

internal fun sendAnchorIdentity(pendingId: String): String = "send-$pendingId"

File: ios/App/Chat/message/SendAnchor.swift
import Foundation

/// Pure one-shot reducer for outbound row viewport anchoring.
struct SendAnchorState: Equatable {
    var observed: Set<String> = []
}

func reduceSendAnchor(_ state: SendAnchorState, identities: [String]) -> (SendAnchorState, String?) {
    let next = identities.last { !state.observed.contains($0) }
    return (SendAnchorState(observed: state.observed.union(identities)), next)
}

func sendAnchorIdentity(_ pendingId: String) -> String { "send-\(pendingId)" }
